diff -up gnome-system-monitor-2.27.4/src/procdialogs.cpp.buttons gnome-system-monitor-2.27.4/src/procdialogs.cpp
--- gnome-system-monitor-2.27.4/src/procdialogs.cpp.buttons	2009-08-22 16:50:30.072430091 -0400
+++ gnome-system-monitor-2.27.4/src/procdialogs.cpp	2009-08-22 16:52:41.435680298 -0400
@@ -190,27 +190,16 @@ procdialog_create_renice_dialog (ProcDat
 	gtk_dialog_set_has_separator (GTK_DIALOG (renice_dialog), FALSE);
 	gtk_container_set_border_width (GTK_CONTAINER (renice_dialog), 5);
 
-	button = gtk_button_new ();
+	button = gtk_button_new_with_mnemonic (_("Change _Priority"));
 	GTK_WIDGET_SET_FLAGS (button, GTK_CAN_DEFAULT);
-		
-	align = gtk_alignment_new (0.5, 0.5, 0.0, 0.0);
-	gtk_container_add (GTK_CONTAINER (button), align);
-		
-	hbox = gtk_hbox_new (FALSE, 2);
-	gtk_container_add (GTK_CONTAINER (align), hbox);
-
 	icon = gtk_image_new_from_stock (GTK_STOCK_OK, GTK_ICON_SIZE_BUTTON);
-	gtk_box_pack_start (GTK_BOX (hbox), icon, FALSE, FALSE, 0);
-
-	label = gtk_label_new_with_mnemonic (_("Change _Priority"));
-	gtk_label_set_mnemonic_widget (GTK_LABEL (label), button);
-	gtk_box_pack_end (GTK_BOX (hbox), label, FALSE, FALSE, 0);
+	gtk_button_set_image (GTK_BUTTON (button), icon);
 
 	gtk_dialog_add_action_widget (GTK_DIALOG (renice_dialog), button, 100);
 
   	gtk_dialog_set_default_response (GTK_DIALOG (renice_dialog), 100);
   	new_nice_value = -100;
-  	  
+
     	dialog_vbox = GTK_DIALOG (dialog)->vbox;
 	gtk_box_set_spacing (GTK_BOX (dialog_vbox), 2);
 	gtk_container_set_border_width (GTK_CONTAINER (dialog_vbox), 5);
